README.md
**THIS REPOSITORY IS OBSOLETE: THE CODE HAS BEEN MERGED IN https://github.com/gem/oq-engine**

**DON'T USE oq-nrmllib!**


debian/changelog
python-oq-nrmllib (0.4.6-1) precise; urgency=low

  * Missing changes added

 -- Matteo Nastasi (GEM Foundation) <nastasi@openquake.org>  Fri, 16 Jan 2015 11:14:08 +0100

python-oq-nrmllib (0.4.6-0) precise; urgency=low

  * Latest improvements

 -- Matteo Nastasi (GEM Foundation) <nastasi@openquake.org>  Fri, 16 Jan 2015 10:57:18 +0100

python-oq-nrmllib (0.4.5-0) precise; urgency=low

  * Expsorure schema: rename the attribute Units to Number (LP: #1187786)
  * Exposure schema update (LP: #1184997, #1184996, #1184981)
  * lxc sandbox - improved CI with sandboxed source tests (LP: #1177319)
  * Add a HazardCurveParser (LP: #1175452)
  * Added non-structural cost attributes and elements (LP: #1174233)
  * Source model XML writer (LP: #1096867)
  * Remove upper bound limit for loss ratio in the vulnerability model and
    loss curve schema (LP: #1171737)
  * Hazard curve in multiple IMTs (LP: #1166107)
  * Importing GMF in XML format: step 1, parsing (LP: #1160398)
  * SES multi surface rupture support (LP: #1144225, #1144207)
  * NRML schema for loss fractions (LP: #1159684)
  * Fragility model with Structure-Dependent intensity measure type
    (LP: #1154549)
  * UHS schema and XML writer (LP: #1154108)
  * Added averageLoss to loss curves (LP: #1152237)
  * Schema and parser support for a new source type: characteristic source
    (LP: #1144207)
  * write mean/stddev in scenario case
  * Continuous Fragility Model fix (LP: #1122832)
  * Link between Rupture / Stochastic Event Set and Ground motion field outputs
    (LP: #1119553)
  * Added FragilityModelParser and related tests, for use in the ScenarioDamage
    calculator (LP: #1100235)
  * Fix the collapse map writer (LP: #1100371)
  * Ported the old writers for the ScenarioDamage calculator into nrml
    (LP: #1100371)
  * Added the noDamageLimit attribute to the fragility model XML (LP: #1100235)
  * Updated the format of the ScenarioDamage output files (LP: #1100371)
  * Update the fragility model to nrml 0.4

 -- Matteo Nastasi (GEM Foundation) <nastasi@openquake.org>  Mon, 24 Jun 2013 13:14:53 +0200

python-oq-nrmllib (0.4.4-0) precise; urgency=low

  * Rename of the package and namespace refactoring

 -- Matteo Nastasi (GEM Foundation) <nastasi@openquake.org>  Sat, 09 Feb 2013 08:44:20 +0100

python-oq-nrmllib (0.4.3-1) precise; urgency=low

  * Upstream release

 -- Muharem Hrnjadovic <mh@foldr3.com>  Thu, 18 Oct 2012 23:14:54 +0200

python-oq-nrmllib (0.4.2-1) precise; urgency=low

  * Fix orig tar ball

 -- Muharem Hrnjadovic <mh@foldr3.com>  Tue, 11 Sep 2012 12:31:37 +0200

python-oq-nrmllib (0.4.1-1) precise; urgency=low

  * Upstream release (LP: #1045211)

 -- Muharem Hrnjadovic <mh@foldr3.com>  Mon, 10 Sep 2012 16:03:10 +0200

python-oq-nrmllib (0.4-1) precise; urgency=low

  * Upstream release

 -- Muharem Hrnjadovic <mh@foldr3.com>  Thu, 26 Jul 2012 09:29:09 +0200


tests/_utils.py
# Copyright (c) 2010-2012, GEM Foundation.
#
# NRML is free software: you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# NRML is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with NRML.  If not, see <http://www.gnu.org/licenses/>.

import collections
from nose import tools
from lxml import etree

import nrmllib


def deep_eq(a, b):
    """Deep compare two objects for equality by traversing __dict__s.

    :returns:
        True if the two objects are deeply equal, otherwise false.
    """
    try:
        _deep_eq(a, b)
    except AssertionError:
        return False
    return True


def _deep_eq(a, b):
    """Do the actual deep comparison. If two items up for comparison is not
    equal, a :exception:`AssertionError` is raised (to :function:`deep_eq`).
    """
    if isinstance(a, (list, tuple)):
        _test_seq(a, b)
    elif isinstance(a, dict):
        _test_dict(a, b)
    elif hasattr(a, '__dict__'):
        assert a.__class__ == b.__class__
        _test_dict(a.__dict__, b.__dict__)
    elif isinstance(a, collections.Iterable) and not isinstance(a, str):
        # If there's a generator or another type of iterable, treat it as a
        # `list`. NOTE: Generators will be exhausted if you do this.
        _test_seq(list(a), list(b))
    else:
        # must be a 'primitive'
        assert a == b


def _test_dict(a, b):
    """Compare `dict` types recursively."""
    assert len(a) == len(b)

    for key in a:
        _deep_eq(a[key], b[key])


def _test_seq(a, b):
    """Compare `list` or `tuple` types recursively."""
    assert len(a) == len(b)

    for i, item in enumerate(a):
        _deep_eq(item, b[i])


def assert_xml_equal(a, b):
    """
    Compare two XML artifacts for equality.

    :param a, b:
        Paths to XML files, or a file-like object containing the XML
        contents.
    """

    contents_a = etree.tostring(etree.parse(a), pretty_print=True)
    contents_b = etree.tostring(etree.parse(b), pretty_print=True)

    tools.assert_equal(contents_a, contents_b)


def validates_against_xml_schema(xml_instance_path,
                                 schema_path=nrmllib.nrml_schema_file()):
    """
    Check whether an XML file validates against an XML schema.
    """

    xml_doc = etree.parse(xml_instance_path)
    xmlschema = etree.XMLSchema(etree.parse(schema_path))
    return xmlschema.validate(xml_doc)


tests/utils_test.py
# Copyright (c) 2010-2012, GEM Foundation.
#
# NRML is free software: you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# NRML is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with NRML.  If not, see <http://www.gnu.org/licenses/>.


import numpy
import unittest

from nrmllib import utils


class UtilsTestCase(unittest.TestCase):
    """Tests for general NRML util functions."""

    def test_coords_to_linestr_wkt_2d(self):
        expected = 'LINESTRING(1.0 1.0, 2.0 2.0, 3.0 3.0)'

        # Mixed input types for robustness testing
        coords = [1.0, '1.0', '2.0', 2.0, 3.0, '3.0']

        actual = utils.coords_to_linestr_wkt(coords, 2)

        self.assertEqual(expected, actual)

    def test_coords_to_linestr_wkt_3d(self):
        expected = 'LINESTRING(1.0 1.0 2.0, 2.0 3.0 3.0)'

        coords = [1.0, '1.0', '2.0', 2.0, 3.0, '3.0']

        actual = utils.coords_to_linestr_wkt(coords, 3)

        self.assertEqual(expected, actual)

    def test_coords_to_poly_wkt_2d(self):
        expected = 'POLYGON((1.0 1.0, 2.0 2.0, 3.0 3.0, 1.0 1.0))'

        coords = [1.0, '1.0', '2.0', 2.0, 3.0, '3.0']

        actual = utils.coords_to_poly_wkt(coords, 2)

        self.assertEqual(expected, actual)

    def test_coords_to_poly_wkt_3d(self):
        expected = (
            'POLYGON((1.0 1.0 0.1, 2.0 2.0 0.2, 3.0 3.0 0.3, 1.0 1.0 0.1))')

        coords = [1.0, '1.0', 0.1, '2.0', 2.0, '0.2', 3.0, '3.0', 0.3]

        actual = utils.coords_to_poly_wkt(coords, 3)

        self.assertEqual(expected, actual)


class NDEnumTestCase(unittest.TestCase):

    def test_1d(self):
        a = numpy.array([x * 3 for x in range(4)])
        expected = [
            ((0,), 0),
            ((1,), 3),
            ((2,), 6),
            ((3,), 9),
        ]
        actual = list(utils.ndenumerate(a))
        self.assertEqual(expected, actual)

    def test_3d(self):
        a = numpy.array([x * 3 for x in range(24)]).reshape((3, 2, 4))
        expected = [
            ((0, 0, 0), 0),
            ((0, 0, 1), 3),
            ((0, 0, 2), 6),
            ((0, 0, 3), 9),
            ((0, 1, 0), 12),
            ((0, 1, 1), 15),
            ((0, 1, 2), 18),
            ((0, 1, 3), 21),
            ((1, 0, 0), 24),
            ((1, 0, 1), 27),
            ((1, 0, 2), 30),
            ((1, 0, 3), 33),
            ((1, 1, 0), 36),
            ((1, 1, 1), 39),
            ((1, 1, 2), 42),
            ((1, 1, 3), 45),
            ((2, 0, 0), 48),
            ((2, 0, 1), 51),
            ((2, 0, 2), 54),
            ((2, 0, 3), 57),
            ((2, 1, 0), 60),
            ((2, 1, 1), 63),
            ((2, 1, 2), 66),
            ((2, 1, 3), 69),
        ]
        actual = list(utils.ndenumerate(a))
        self.assertEqual(expected, actual)

    def test_4d(self):
        a = numpy.array([x * 3 for x in range(48)]).reshape((2, 2, 3, 4))
        expected = [
            ((0, 0, 0, 0), 0),
            ((0, 0, 0, 1), 3),
            ((0, 0, 0, 2), 6),
            ((0, 0, 0, 3), 9),
            ((0, 0, 1, 0), 12),
            ((0, 0, 1, 1), 15),
            ((0, 0, 1, 2), 18),
            ((0, 0, 1, 3), 21),
            ((0, 0, 2, 0), 24),
            ((0, 0, 2, 1), 27),
            ((0, 0, 2, 2), 30),
            ((0, 0, 2, 3), 33),
            ((0, 1, 0, 0), 36),
            ((0, 1, 0, 1), 39),
            ((0, 1, 0, 2), 42),
            ((0, 1, 0, 3), 45),
            ((0, 1, 1, 0), 48),
            ((0, 1, 1, 1), 51),
            ((0, 1, 1, 2), 54),
            ((0, 1, 1, 3), 57),
            ((0, 1, 2, 0), 60),
            ((0, 1, 2, 1), 63),
            ((0, 1, 2, 2), 66),
            ((0, 1, 2, 3), 69),
            ((1, 0, 0, 0), 72),
            ((1, 0, 0, 1), 75),
            ((1, 0, 0, 2), 78),
            ((1, 0, 0, 3), 81),
            ((1, 0, 1, 0), 84),
            ((1, 0, 1, 1), 87),
            ((1, 0, 1, 2), 90),
            ((1, 0, 1, 3), 93),
            ((1, 0, 2, 0), 96),
            ((1, 0, 2, 1), 99),
            ((1, 0, 2, 2), 102),
            ((1, 0, 2, 3), 105),
            ((1, 1, 0, 0), 108),
            ((1, 1, 0, 1), 111),
            ((1, 1, 0, 2), 114),
            ((1, 1, 0, 3), 117),
            ((1, 1, 1, 0), 120),
            ((1, 1, 1, 1), 123),
            ((1, 1, 1, 2), 126),
            ((1, 1, 1, 3), 129),
            ((1, 1, 2, 0), 132),
            ((1, 1, 2, 1), 135),
            ((1, 1, 2, 2), 138),
            ((1, 1, 2, 3), 141),
        ]
        actual = list(utils.ndenumerate(a))
        self.assertEqual(expected, actual)


tests/risk/parsers_test.py
# coding=utf-8
# Copyright (c) 2010-2012, GEM Foundation.
#
# OpenQuake is free software: you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# OpenQuake is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with OpenQuake.  If not, see <http://www.gnu.org/licenses/>.

import os
import unittest
import StringIO

from nrmllib.risk import parsers


EXAMPLES_DIR = os.path.join(
    os.path.dirname(os.path.dirname(os.path.dirname(__file__))),
    'examples')


def get_example(fname):
    return os.path.join(EXAMPLES_DIR, fname)


class ExposureModelParserTestCase(unittest.TestCase):

    def test_schema_validation(self):
        invalid_exposure = """\
<?xml version='1.0' encoding='utf-8'?>
  <nrml xmlns:gml="http://www.opengis.net/gml"
    xmlns="http://openquake.org/xmlns/nrml/0.4">
    <exposureModel gml:id="ep1"/>
</nrml>
"""

        self.assertRaises(ValueError, parsers.ExposureModelParser,
                          StringIO.StringIO(invalid_exposure))

    def test_parsing(self):
        exposure = """\
<?xml version='1.0' encoding='utf-8'?>
  <nrml xmlns:gml="http://www.opengis.net/gml"
    xmlns="http://openquake.org/xmlns/nrml/0.4">

  <exposureModel gml:id="ep1">
    <config/>
    <exposureList gml:id="PAV01" areaType="per_asset" areaUnit="GBP"
      assetCategory="buildings" cocoType="per_area" cocoUnit="CHF"
      recoType="aggregated" recoUnit="EUR" stcoType="aggregated"
      stcoUnit="USD">

      <gml:description>Buildings in Pavia</gml:description>
      <taxonomySource>Pavia taxonomy</taxonomySource>
      <assetDefinition gml:id="asset_01">
        <site>
          <gml:Point srsName="epsg:4326">
            <gml:pos>9.15000 45.16667</gml:pos>
          </gml:Point>
        </site>

        <area>120</area>
        <coco>12.95</coco>
        <deductible>55</deductible>
        <limit>999</limit>
        <number>7</number>
        <reco>109876</reco>
        <stco>150000</stco>
        <taxonomy>RC/DMRF-D/LR</taxonomy>
      </assetDefinition>

      <assetDefinition gml:id="asset_02">
        <site>
          <gml:Point srsName="epsg:4326">
            <gml:pos>9.15333 45.12200</gml:pos>
          </gml:Point>
        </site>

        <area>119</area>
        <coco>21.95</coco>
        <deductible>66</deductible>
        <limit>1999</limit>
        <number>6</number>
        <occupants description="day">12</occupants>
        <occupants description="night">50</occupants>
        <reco>205432</reco>
        <stco>250000</stco>
        <taxonomy>RC/DMRF-D/HR</taxonomy>
      </assetDefinition>

    </exposureList>
  </exposureModel>
</nrml>
"""

        expected_result = [
            ([9.15000, 45.16667], [], {
                "area": 120.0,
                "areaType": "per_asset",
                "areaUnit": "GBP",
                "assetCategory": "buildings",
                "assetID": "asset_01",
                "coco": 12.95,
                "cocoType": "per_area",
                "cocoUnit": "CHF",
                "deductible": 55.0,
                "limit": 999.0,
                "listDescription": "Buildings in Pavia",
                "listID": "PAV01",
                "number": 7.0,
                "reco": 109876.0,
                "recoType": "aggregated",
                "recoUnit": "EUR",
                "stco": 150000.0,
                "stcoType": "aggregated",
                "stcoUnit": "USD",
                "taxonomy": "RC/DMRF-D/LR",
                "taxonomySource": "Pavia taxonomy",
            }),
            ([9.15333, 45.12200], [
            parsers.OCCUPANCY(12, "day"), parsers.OCCUPANCY(50, "night")], {
                 "area": 119.0,
                 "areaType": "per_asset",
                 "areaUnit": "GBP",
                 "assetCategory": "buildings",
                 "assetID": "asset_02",
                 "coco": 21.95,
                 "cocoType": "per_area",
                 "cocoUnit": "CHF",
                 "deductible": 66.0,
                 "limit": 1999.0,
                 "listDescription": "Buildings in Pavia",
                 "listID": "PAV01",
                 "number": 6.0,
                 "reco": 205432.0,
                 "recoType": "aggregated",
                 "recoUnit": "EUR",
                 "stco": 250000.0,
                 "stcoType": "aggregated",
                 "stcoUnit": "USD",
                 "taxonomy": "RC/DMRF-D/HR",
                 "taxonomySource": "Pavia taxonomy",
            }),
        ]

        parser = parsers.ExposureModelParser(StringIO.StringIO(exposure))
        for ctr, (exposure_point, occupancy_data, exposure_data) \
                in enumerate(parser):

            self.assertEqual(expected_result[ctr][0], exposure_point)
            self.assertEqual(expected_result[ctr][1], occupancy_data)
            self.assertEqual(expected_result[ctr][2], exposure_data)


class VulnerabilityModelParserTestCase(unittest.TestCase):

    def test_schema_validation(self):
        invalid_vulnerability_model = """\
<?xml version='1.0' encoding='utf-8'?>
  <nrml xmlns:gml="http://www.opengis.net/gml"
    xmlns="http://openquake.org/xmlns/nrml/0.4">
    <vulnerabilityModel/>
</nrml>
"""

        self.assertRaises(ValueError, parsers.VulnerabilityModelParser,
                          StringIO.StringIO(invalid_vulnerability_model))

    def test_parsing(self):
        vulnerability_model = """\
<?xml version='1.0' encoding='utf-8'?>
<nrml xmlns="http://openquake.org/xmlns/nrml/0.4"
  xmlns:gml="http://www.opengis.net/gml">

  <vulnerabilityModel>
    <config/>
    <discreteVulnerabilitySet vulnerabilitySetID="PAGER"
      assetCategory="population" lossCategory="fatalities">
      <IML IMT="MMI">5.00 5.50 6.00 6.50</IML>
      <discreteVulnerability vulnerabilityFunctionID="IR" probabilisticDistribution="LN">
        <lossRatio>0.18 0.36 0.36 0.36</lossRatio>
        <coefficientsVariation>0.30 0.30 0.30 0.30</coefficientsVariation>
      </discreteVulnerability>
      <discreteVulnerability vulnerabilityFunctionID="PK" probabilisticDistribution="LN">
        <lossRatio>0.18 0.36 0.36 0.36</lossRatio>
        <coefficientsVariation>0.30 0.30 0.30 0.30</coefficientsVariation>
      </discreteVulnerability>
    </discreteVulnerabilitySet>
    <discreteVulnerabilitySet vulnerabilitySetID="NPAGER"
      assetCategory="population" lossCategory="fatalities">
      <IML IMT="MMI">6.00 6.50 7.00 7.50</IML>
      <discreteVulnerability vulnerabilityFunctionID="AA" probabilisticDistribution="LN">
        <lossRatio>0.00 0.00 0.00 0.00</lossRatio>
        <coefficientsVariation>0.50 0.50 0.50 0.50</coefficientsVariation>
      </discreteVulnerability>
      <discreteVulnerability vulnerabilityFunctionID="BB" probabilisticDistribution="LN">
        <lossRatio>0.06 0.18 0.36 0.36</lossRatio>
        <coefficientsVariation>0.30 0.30 0.30 0.30</coefficientsVariation>
      </discreteVulnerability>
    </discreteVulnerabilitySet>
  </vulnerabilityModel>
</nrml>
"""

        model = self._load_model(StringIO.StringIO(vulnerability_model))

        self.assertEqual("MMI", model["PK"]["IMT"])
        self.assertEqual("fatalities", model["PK"]["lossCategory"])
        self.assertEqual("PAGER", model["PK"]["vulnerabilitySetID"])
        self.assertEqual("population", model["PK"]["assetCategory"])
        self.assertEqual("LN", model["PK"]["probabilisticDistribution"])

        self.assertEqual([0.18, 0.36, 0.36, 0.36], model["PK"]["lossRatio"])

        self.assertEqual([0.30, 0.30, 0.30, 0.30],
                         model["PK"]["coefficientsVariation"])

        self.assertEqual([5.00, 5.50, 6.00, 6.50], model["PK"]["IML"])
        self.assertEqual([0.18, 0.36, 0.36, 0.36], model["IR"]["lossRatio"])

        self.assertEqual([0.30, 0.30, 0.30, 0.30],
                         model["IR"]["coefficientsVariation"])

        self.assertEqual([5.00, 5.50, 6.00, 6.50], model["IR"]["IML"])
        self.assertEqual("NPAGER", model["AA"]["vulnerabilitySetID"])
        self.assertEqual([6.00, 6.50, 7.00, 7.50], model["AA"]["IML"])

        self.assertEqual([0.50, 0.50, 0.50, 0.50],
                         model["AA"]["coefficientsVariation"])

    def _load_model(self, source):
        model = dict()
        parser = parsers.VulnerabilityModelParser(source)

        for vulnerability_function in parser:
            model[vulnerability_function["ID"]] = vulnerability_function

        return model


class FragilityModelParserTestCase(unittest.TestCase):

    def test_damage_states_bad_ordering(self):
        fm_file = StringIO.StringIO('''<?xml version='1.0' encoding='utf-8'?>
<nrml xmlns="http://openquake.org/xmlns/nrml/0.4">

    <fragilityModel format="continuous" imlUnit="m" minIML="0.1" maxIML="9.9">
        <description>Fragility model for Pavia (continuous)</description>
        <!-- limit states apply to the entire fragility model -->
        <limitStates>
            slight
            moderate
            extensive
            complete
        </limitStates>

        <!-- fragility function set, each with its own, distinct taxonomy -->
        <ffs noDamageLimit="0.05" type="lognormal">
            <taxonomy>RC/DMRF-D/LR</taxonomy>

            <!-- fragility function in continuous format, 1 per limit state -->
            <ffc ls="slight">
                <params mean="11.19" stddev="8.27" />
            </ffc>

            <ffc ls="extensive">
                <params mean="48.05" stddev="42.49" />
            </ffc>

            <ffc ls="moderate">
                <params mean="27.98" stddev="20.677" />
            </ffc>

            <ffc ls="complete">
                <params mean="108.9" stddev="123.7" />
            </ffc>
        </ffs>
 </fragilityModel>
</nrml>
''')
        self.assertRaises(ValueError, list,
                          parsers.FragilityModelParser(fm_file))

    def test_parse_continuous(self):
        p = iter(parsers.FragilityModelParser(get_example('fragm_c.xml')))

        format, iml, limit_states = p.next()
        self.assertEqual(format, 'continuous')
        self.assertIsNone(iml['IMT'])
        self.assertIsNone(iml['imls'])
        self.assertEqual(limit_states,
                         ['slight', 'moderate', 'extensive', 'complete'])

        ffs1, ffs2 = list(p)
        self.assertEqual(ffs1,
                         ('RC/DMRF-D/LR',
                          [(11.19, 8.27),
                           (27.98, 20.677),
                           (48.05, 42.49),
                           (108.9, 123.7)], 0.05))
        self.assertEqual(ffs2,
                         ('RC/DMRF-D/HR',
                          [(11.18, 8.28),
                           (27.99, 20.667),
                           (48.06, 42.48),
                           (108.8, 123.6)], None))

    def test_parse_discrete(self):
        p = iter(parsers.FragilityModelParser(get_example('fragm_d.xml')))

        format, iml, limit_states = p.next()
        self.assertEqual(format, 'discrete')
        self.assertEqual(iml['IMT'], 'MMI')
        self.assertEqual(iml['imls'], [7.0, 8.0, 9.0, 10.0, 11.0])
        self.assertEqual(limit_states,
                         ['minor', 'moderate', 'severe', 'collapse'])

        ffs1, ffs2 = list(p)
        self.assertEqual(ffs1,
                         ('RC/DMRF-D/LR',
                          [[0.0, 0.09, 0.56, 0.91, 0.98],
                           [0.0, 0.0, 0.04, 0.78, 0.96],
                           [0.0, 0.0, 0.0, 0.29, 0.88],
                           [0.0, 0.0, 0.0, 0.03, 0.63]], 5.0))
        self.assertEqual(ffs2,
                         ('RC/DMRF-D/HR',
                          [[0.0, 0.09, 0.56, 0.92, 0.99],
                           [0.0, 0.0, 0.04, 0.79, 0.97],
                           [0.0, 0.0, 0.0, 0.3, 0.89],
                           [0.0, 0.0, 0.0, 0.04, 0.64]], None))


nrmllib/risk/parsers.py
# coding=utf-8
# Copyright (c) 2010-2012, GEM Foundation.
#
# OpenQuake is free software: you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# OpenQuake is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with OpenQuake.  If not, see <http://www.gnu.org/licenses/>.

"""
Module containing parsers for risk input artifacts.
"""

from lxml import etree
from collections import namedtuple

import nrmllib

NRML = "{%s}" % nrmllib.NAMESPACE
GML = "{%s}" % nrmllib.GML_NAMESPACE

OCCUPANCY = namedtuple("OCCUPANCY", "occupants, description")


class ExposureModelParser(object):
    """
    Exposure model parser. This class is implemented as a generator.

    For each `assetDefinition` element in the parsed document,
    it yields a tuple, where:

    * the first element is a list with the geographical information,
      in the following format: [lon, lat].
    * the second element is a list of occupancy objects, each one with
      a property called `occupants` (the number of
      occupants), and a property called `description` (the context
      in which the number of occupants has been measured, for example
      during the day or night).
    * the third element is a dictionary with all the other information
      related to the asset. The attributes dictionary looks like this:
      {'listID': 'PAV01',
       'listDescription': 'Collection of existing building in Pavia',
       'assetID': 'asset_02',
       'assetCategory': 'buildings',
       'taxonomy': 'RC/DMRF-D/LR',
       'structureCategory': 'RC-LR-PC',
       'assetValue': 250000.0,
       'assetValueUnit': 'EUR'}

    :param source:
        Filename or file-like object containing the XML data.
    """

    def __init__(self, source):
        self._source = source
        assert_is_valid(self._source)

        # contains the data of the node currently parsed.
        self._current_meta = {}

    def assert_is_valid(self):
        """
        Check the input file is valid according to the schema.
        """

        exposure = etree.parse(self._source)
        xmlschema = etree.XMLSchema(etree.parse(
                nrmllib.nrml_schema_file()))

        if not xmlschema.validate(exposure):
            raise ValueError("Exposure model is not valid.")

    def __iter__(self):
        for i in self._parse():
            yield i

    def _parse(self):
        """
        Parse the document iteratively.
        """

        schema = etree.XMLSchema(etree.parse(
                nrmllib.nrml_schema_file()))

        for event, element in etree.iterparse(
                self._source, events=('start', 'end'), schema=schema):

            if event == 'start' and element.tag == '%sexposureList' % NRML:
                # we need to get the exposureList id, description and
                # asset category.
                exp_id = element.get('%sid' % GML)
                self._current_meta['listID'] = str(exp_id)

                desc = element.find('%sdescription' % GML)
                if desc is not None:
                    self._current_meta['listDescription'] = str(desc.text)

                taxsrc = element.find('%staxonomySource' % NRML)
                if taxsrc is not None:
                    self._current_meta['taxonomySource'] = str(taxsrc.text)

                asset_category = str(element.get('assetCategory'))
                self._current_meta['assetCategory'] = asset_category

                # type and unit for area, contents cost, retrofitting cost
                # and structural cost.
                attrs = ("areaType", "areaUnit", "cocoType", "cocoUnit",
                         "recoType", "recoUnit", "stcoType", "stcoUnit")
                for attr_name in attrs:
                    attr_value = element.get(attr_name)
                    if attr_value is not None:
                        self._current_meta[attr_name] = attr_value

            elif event == 'end' and element.tag == '%sassetDefinition' % NRML:
                site_data = (_to_site(element), _to_occupancy(element),
                             self._to_site_attributes(element))
                del element
                yield site_data

    def _to_site_attributes(self, element):
        """
        Build a dict of all node attributes.
        """

        site_attributes = {}
        site_attributes['assetID'] = element.get('%sid' % GML)

        # Optional elements.
        attrs = (('coco', float), ('reco', float), ('stco', float),
                 ('area', float), ('number', float), ('limit', float),
                 ('deductible', float))
        for (attr_name, attr_type) in attrs:
            attr_value = element.find('%s%s' % (NRML, attr_name))
            if attr_value is not None:
                site_attributes[attr_name] = attr_type(attr_value.text)

        # Mandatory elements.
        for (required_attr, attr_type) in (('taxonomy', str),):
            attr_value = element.find('%s%s' % (NRML, required_attr)).text
            if attr_value is not None:
                site_attributes[required_attr] = attr_type(attr_value)
            else:
                error_str = ("element assetDefinition: missing required "
                             "attribute %s" % required_attr)
                raise ValueError(error_str)

        site_attributes.update(self._current_meta)
        return site_attributes


def _to_occupancy(element):
    """
    Convert the 'occupants' tags to named tuples.

    We want to extract the value of <occupants>. We expect the input
    element to be an 'assetDefinition' and have a child element
    structured like this:

    <occupants description="day">245</occupants>
    """

    occupancy_data = []
    for otag in element.findall('%soccupants' % NRML):
        occupancy_data.append(OCCUPANCY(
            occupants=int(otag.text), description=otag.attrib["description"]))
    return occupancy_data


def _to_site(element):
    """
    Convert current GML attributes to Site object.

    We want to extract the value of <gml:pos>. We expect the input
    element to be an 'assetDefinition' and have a child element
    structured like this:

    <site>
        <gml:Point srsName="epsg:4326">
            <gml:pos>9.15000 45.16667</gml:pos>
        </gml:Point>
    </site>
    """

    point_elem = element.find('%ssite' % NRML).find('%sPoint' % GML)
    return [float(x.strip()) for x in point_elem.find(
        '%spos' % GML).text.split()]


class VulnerabilityModelParser(object):
    """
    Vulnerability model parser. This class is implemented as a generator.

    For each `discreteVulnerability` element in the parsed document,
    it yields a dictionary containing the function attributes.

    :param source:
        Filename or file-like object containing the XML data.
    """

    def __init__(self, source):
        self._source = source
        assert_is_valid(self._source)

        self._vulnerability_model = etree.parse(self._source).getroot()

    def __iter__(self):
        """
        Parse the vulnerability model.
        """

        for vulnerability_set in self._vulnerability_model.findall(
                ".//%sdiscreteVulnerabilitySet" % NRML):

            vulnerability_function = self._parse_set_attributes(
                vulnerability_set)

            for vf in vulnerability_set.findall(
                    ".//%sdiscreteVulnerability" % NRML):

                loss_ratios = [float(x) for x in vf.find(
                    "%slossRatio" % NRML).text.strip().split()]

                coefficients_variation = [float(x) for x in vf.find(
                    "%scoefficientsVariation" % NRML).text.strip().split()]

                vulnerability_function["ID"] = vf.attrib[
                    "vulnerabilityFunctionID"]

                vulnerability_function["probabilisticDistribution"] = \
                    vf.attrib["probabilisticDistribution"]

                vulnerability_function["lossRatio"] = loss_ratios
                vulnerability_function["coefficientsVariation"] = \
                    coefficients_variation

                yield dict(vulnerability_function)

    @staticmethod
    def _parse_set_attributes(vset):
        """
        Extract the attributes common to all the vulnerability functions
        belonging to the given set.
        """

        attrs = dict()
        imls = vset.find(".//%sIML" % NRML)

        attrs["IMT"] = imls.attrib["IMT"]
        attrs["IML"] = [float(x) for x in imls.text.strip().split()]

        attrs["lossCategory"] = vset.attrib["lossCategory"]
        attrs["assetCategory"] = vset.attrib["assetCategory"]
        attrs["vulnerabilitySetID"] = vset.attrib["vulnerabilitySetID"]

        return attrs


def find(tag, elem):
    "Find all the subelements matching the given tag"
    return elem.findall(NRML + tag, elem)


def findone(tag, elem, default=None):
    """
    Find the unique subelement matching the given tag. Raise ValueError
    if there are too many elements and returns the default if there is
    no match.
    """
    elems = elem.findall(NRML + tag, elem)
    n = len(elems)
    if n == 0:  # not found
        return default
    if n > 1:
        raise ValueError('Found %d elements of kind %s, expected one'
                         % (n, tag))
    return elems[0]


class FragilityModelParser(object):
    """
    Fragility model parser. This class is implemented as a generator.
    It yields a triple (format, iml, limit_states), associated to a
    fragility model, followed by a sequence of triples of the form
    (taxonomy, params, no_damage_limit), associated each to a
    different fragility function sequence.

    :param source:
        Filename or file-like object containing the XML data.
    """

    def __init__(self, source):
        self._source = source
        assert_is_valid(self._source)
        self._fragility_model = etree.parse(self._source).getroot()

    def __iter__(self):
        """
        Parse the fragility model.
        """
        fragilityModel = findone('fragilityModel', self._fragility_model)
        format = fragilityModel.attrib['format']
        if format == 'discrete':
            iml = findone('IML', fragilityModel)
            iml = dict(IMT=iml.attrib['IMT'],
                       imls=map(float, iml.text.split()))
        else:
            iml = dict(IMT=None, imls=None)
        self.limit_states = findone('limitStates', fragilityModel).text.split()
        yield format, iml, self.limit_states
        for ffs in find('ffs', fragilityModel):
            taxonomy = findone('taxonomy', ffs).text
            no_damage_limit = ffs.attrib.get('noDamageLimit')
            if no_damage_limit:
                no_damage_limit = float(no_damage_limit)
            if format == 'discrete':
                all_poes = []
                for lsi, ffd in enumerate(find('ffd', ffs)):
                    self._check_limit_state(lsi, ffd.attrib['ls'])
                    all_poes.append(
                        map(float, findone('poEs', ffd).text.split()))
                yield taxonomy, all_poes, no_damage_limit
            else:  # continuous
                all_params = []
                for lsi, ffc in enumerate(find('ffc', ffs)):
                    self._check_limit_state(lsi, ffc.attrib['ls'])
                    params = findone('params', ffc).attrib
                    all_params.append(
                        (float(params['mean']), float(params['stddev'])))
                yield taxonomy, all_params, no_damage_limit

    def _check_limit_state(self, lsi, ls):
        if ls != self.limit_states[lsi]:
            raise ValueError('Expected limitState %s, got %s' %
                             (self.limit_states[lsi], ls))


def assert_is_valid(source):
    """
    Check the input file is valid according to the schema.
    """

    exposure = etree.parse(source)
    xmlschema = etree.XMLSchema(etree.parse(
            nrmllib.nrml_schema_file()))

    if not xmlschema.validate(exposure):
        raise ValueError("Exposure model is not valid.")
